### ISSUE TITLE
Why not a more modern alternative to PBKDF2?

### ISSUE THREAD

**Issue by hpkns** [defuse/php-encryption#409]
I really appreciate the job you've been doing to provide a safe and convenient PHP encryption library, and I have learned quite a lot by reading your sources, but there is something I don't quite understand: why wouldn't you use a "better" alternative to PBKDF2?

It's been out there for quite some times, and it's now (relatively) easy to brute-force using a GPU rig. Alternatives like bcrypt or argon2 are much more resilient to such attacks (argon2 much more so that bcrypt). 

Is it just a just a backward compatibility (argon2 being a very recent addition to PHP) or is it any other reason that I don't see?

**Comment by paragonie-scott** [comment]
Because implementing scrypt or Argon2 in pure PHP is so ungodly slow that it's not practical.

**Comment by paragonie-scott** [comment]
Also, bcrypt isn't a KDF it's only meant for password hashing. (Attempting to generate keys larger than 192 bits with bcrypt is undefined in the literature.)

**Comment by hpkns** [comment]
I get the argument for those two, but if I'm not mistaken, bcrypt has been available in PHP for quite some times and seems to resist brute force attacks better than pbkdf2...

**Comment by hpkns** [comment]
Thank you for your answer!

But isn't it worth it to loose 64 bits of entropy (you'll still get a 256 bits key out of HKDF anyway) to make iterating though passwords slower?

As far I know, 192 bits is order of magnitudes more than what is possible to brute-force, even at a super-computer level and the weakest point in such case is always the password…

**Comment by paragonie-scott** [comment]
> But isn't it worth it to loose 64 bits of entropy (you'll still get a 256 bits key out of HKDF anyway) to make iterating though passwords slower?

No, it isn't worth it. HKDF doesn't increase entropy. If you only have 2^192 possible inputs, the number of possible outputs is only (at most) 2^192.
